python_modules/dagster/dagster/_core/definitions/run_request.py

```diff
@@ -90,19 +100,88 @@ def __new__(
                 asset_selection, "asset_selection", of_type=AssetKey
             ),
             stale_assets_only=check.bool_param(stale_assets_only, "stale_assets_only"),
+            partition_key=check.opt_str_param(partition_key, "partition_key"),
         )
 
-    @property
-    def partition_key(self) -> Optional[str]:
-        return self.tags.get(PARTITION_NAME_TAG)
-
     def with_replaced_attrs(self, **kwargs: Any) -> "RunRequest":
         fields = self._asdict()
         for k in fields.keys():
             if k in kwargs:
                 fields[k] = kwargs[k]
         return RunRequest(**fields)
 
+    def with_resolved_partition(
+        self,
+        target_definition: Union["JobDefinition", "UnresolvedAssetJobDefinition"],
+        current_time: Optional[datetime] = None,
+        dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
+    ) -> "RunRequest":
+        from dagster._core.definitions.job_definition import JobDefinition
+
+        if self.partition_key is None:
+            check.failed(
+                "Cannot resolve partition for run request without partition key",
+            )
+
+        partitions_def = target_definition.partitions_def
+        if partitions_def is None:
+            check.failed(
+                "Cannot resolve partition for run request without partitions_def",
+            )
+        partitions_def = cast(PartitionsDefinition, partitions_def)
```

Comment:
It seems like we already have the `partitions_def` in all the places this method is called? If so, maybe we should just pass it in? You can ignore this comment.

Reply:
It is duplicative to check partitions def again here, but I prefer it as it ensures the partitions def is bound to the target